src/path_tracking_pid_local_planner.cpp

```diff
@@ -0,0 +1,644 @@
+//
+// Created by nobleo on 12-9-18.
+//
+
+#include "path_tracking_pid/path_tracking_pid_local_planner.hpp"
+#include <algorithm>
+#include <geometry_msgs/TransformStamped.h>
+#include <limits>
+#include <memory>
+#include <pluginlib/class_list_macros.h>
+#include <string>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#include <tf2_ros/transform_listener.h>
+#include <vector>
+
+// register planner as move_base and move_base plugins
+PLUGINLIB_EXPORT_CLASS(path_tracking_pid::TrackingPidLocalPlanner, nav_core::BaseLocalPlanner)
+PLUGINLIB_EXPORT_CLASS(path_tracking_pid::TrackingPidLocalPlanner, mbf_costmap_core::CostmapController)
+
+namespace path_tracking_pid
+{
+TrackingPidLocalPlanner::TrackingPidLocalPlanner() = default;
+
+TrackingPidLocalPlanner::~TrackingPidLocalPlanner() = default;
+
+void TrackingPidLocalPlanner::reconfigure_pid(path_tracking_pid::PidConfig& config, uint32_t level)
+{
+  pid_controller_.configure(config);
+  controller_debug_enabled_ = config.controller_debug_enabled;
+
+  if (controller_debug_enabled_ && !global_plan_.empty())
+  {
+    received_path_.header = global_plan_.at(0).header;
+    received_path_.poses = global_plan_;
+    path_pub_.publish(received_path_);
+  }
+}
+
+void TrackingPidLocalPlanner::initialize(std::string name, tf2_ros::Buffer* tf, costmap_2d::Costmap2DROS* costmap)
+{
+  ros::NodeHandle nh("~/" + name);
+  ros::NodeHandle gn;
+  ROS_DEBUG("TrackingPidLocalPlanner::initialize(%s, ..., ...)", name.c_str());
+  // setup dynamic reconfigure
+  pid_server_ = std::make_unique<dynamic_reconfigure::Server<path_tracking_pid::PidConfig>>(config_mutex_, nh);
+  dynamic_reconfigure::Server<path_tracking_pid::PidConfig>::CallbackType cb1;
+  cb1 = boost::bind(&TrackingPidLocalPlanner::reconfigure_pid, this, _1, _2);
+  pid_server_->setCallback(cb1);
+  pid_controller_.setEnabled(false);
+
+  bool holonomic_robot;
+  nh.param<bool>("holonomic_robot", holonomic_robot, false);
+  pid_controller_.setHolonomic(holonomic_robot);
+
+  bool estimate_pose_angle;
+  nh.param<bool>("estimate_pose_angle", estimate_pose_angle, false);
+  pid_controller_.setEstimatePoseAngle(estimate_pose_angle);
+
+  nh.param<std::string>("base_link_frame", base_link_frame_, "base_link");
+
+  nh.param<bool>("use_tricycle_model", use_tricycle_model_, false);
+  nh.param<std::string>("steered_wheel_frame", steered_wheel_frame_, "steer");
```

Comment:
Vehicles rarely get a fourth wheel on runtime :slightly_smiling_face: 

And frame-names are assumed to be static in ROS mostly?

Reply:
For me the cfg file is very useful to get a quick look at which parameters are available. It would be nice if all parameters would be described in it. Even if some parameters won't change on runtime

Reply:
Can be done, but then the code should cope with changing them.